cnd/src/seed.rs

```diff
@@ -40,6 +48,168 @@ impl Seed {
         rand.try_fill(&mut arr[..])?;
         Ok(Seed(arr))
     }
+
+    /// Construct a seed from base64 data read in from pem file.
+    pub fn from_file<D: AsRef<OsStr>>(seed_file: D) -> Result<Seed, Error> {
+        let file = Path::new(&seed_file);
+        log::info!(
+            "Found secret seed file, reading from {}",
+            PrintablePath(&file.to_path_buf())
+        );
+
+        let contents = fs::read_to_string(file)?;
+        let pem = pem::parse(contents)?;
+        let bytes = seed_from_pem(&pem::encode(&pem))?;
+
+        Ok(Seed::from(bytes))
+    }
+
+    /// Read the seed from the default location if it exists, otherwise
+    /// generate a random seed and write it to the default location.
+    pub fn from_default_file_or_generate<R: Rng>(rand: R) -> Result<Seed, Error> {
+        let path = default_seed_path()?;
+
+        if path.exists() {
+            return Self::from_file(path);
+        }
+
+        let random_seed = Seed::new_random(rand)?;
+        random_seed.write_to(path)?;
+        Ok(random_seed)
+    }
+
+    fn write_to(&self, seed_file: PathBuf) -> Result<(), Error> {
+        ensure_directory_exists(seed_file.clone())?;
+        self.write_to_file(seed_file)?;
+        Ok(())
+    }
+
+    fn write_to_file(&self, path: PathBuf) -> Result<(), Error> {
+        let out = format!(
+            "-----BEGIN SECRET SEED-----\n{}\n-----END SECRET SEED-----\n",
```

Comment:
If you are doing all of this manually, what is the `pem` library good for? :thinking:

Reply:
Good question, I asked this myself - that library is practically useless.

Reply:
On a serious note, it does the validation of the pem format i.e., opening tag, data, cloning tag.

Reply:
According to the docs, you should be able to parse and encode the file contents with it?

- https://docs.rs/pem/0.6.1/pem/fn.parse.html
- https://docs.rs/pem/0.6.1/pem/struct.Pem.html#structfield.contents
- https://docs.rs/pem/0.6.1/pem/fn.encode.html

Reply:
@tcharding Please address this before merging, I think a fair bit of the logic here can be offloaded to the library :)

Reply:
ha ha - I sat down to write a pem library tonight after the kids went to bed because of the terrible code I wrote in this PR.  Before starting I thought I had better take one more look at the library - well, as you posted here, turns out I can't read a Rust library.  Adding patches to fix this and use the library correctly.  Thanks for looking into this @thomaseizinger and for doing the Typescript stuff, I owe you a beer.